diff --git a/rootdir/init.rc b/rootdir/init.rc
index 0510696..6a2888a 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -397,6 +397,12 @@ service netd /system/bin/netd
 service debuggerd /system/bin/debuggerd
     class main
 
+#Open NFC server
+service server_open_nfc /system/bin/server_open_nfc
+    class main
+    user root
+    group root
+
 service ril-daemon /system/bin/rild
     class main
     socket rild stream 660 root radio
diff --git a/rootdir/ueventd.rc b/rootdir/ueventd.rc
index 438cf0a..327ccb6 100644
--- a/rootdir/ueventd.rc
+++ b/rootdir/ueventd.rc
@@ -76,6 +76,9 @@
 /dev/ts0710mux*           0640   radio      radio
 /dev/ppp                  0660   radio      vpn
 
+#NFC device
+/dev/nfcc                 0600   root       root
+
 # sysfs properties
 /sys/devices/virtual/input/input*   enable      0660  root   input
 /sys/devices/virtual/input/input*   poll_delay  0660  root   input
